### ISSUE TITLE
better cache persistence as an option

### ISSUE THREAD

**Issue by techyowl** [onevcat/Kingfisher#702]
### Check List

Thanks for considering to open an issue. Before you submit your issue, please confirm these boxes are checked.

- [X ] I have read the [wiki page](https://github.com/onevcat/Kingfisher/wiki) and [cheat sheet](https://github.com/onevcat/Kingfisher/wiki/Cheat-Sheet), but there is no information I need.
- [X ] I have searched in [existing issues](https://github.com/onevcat/Kingfisher/issues?utf8=✓&q=is%3Aissue), but did find a same one.
- [ X] I want to reoprt a problem instead of asking a question. It'd better to use [kingfisher tag in Stack Overflow](http://stackoverflow.com/questions/tagged/kingfisher) to ask a question.

#### What

[Tell us about the issue]

I'm not sure exactly why but I have set the 
        ImageCache.default.maxMemoryCost = 1 and I've set ImageCache.default.maxCachePeriodInSecond = -1.
when I am using the image prefetcher to obtain images from a json response with the image urls not changing from time to time as the json is cached also. and then for displaying the images I'm trying to use this. ImageCache.default.retrieveImageInDiskCache(forKey: img) where omg is a string for a url path that was used in the prefetched. now right after the prefetched runs it works perfectly however after a day or two and reloading from cache most of the images return nil. I need to have as much persistence as possible however something seems to be clearing the cache. Is their something in the api I have missed or is it not possible to have this level or persistence I'm seeking.  I also don't set a max cache size since it only supports 32bit integer and 50gb In bytes won't work.

#### Reproduce

[The steps to reproduce this issue. What is the url you were trying to load, where did you put your code, etc.]

#### Other Comment

[Add anything else here]

**Comment by onevcat** [comment]
By default, Kingfisher will cache the images in system cache folder. If you are not setting the max disk cache, the default value "0" will be used, which means unlimited disk cache. So Kingfisher will never purge your disk cache.

However, since the system cache path is used, there is a chance that the system would delete that folder or remove some files under the folder when there is no enough disk (or by some other similar strategy, which is an implementation detail of iOS system). Since it is possible to access the cache folder on your phone or simulator, I suggest to check whether the folder and files are still there.

So it means using the cache folder is not a good idea as a persistence mean. It could be purged by system without notice. If you really need to persist the downloaded images, you could try to create a new `ImageCache` with a [diskCachePathClosure](https://github.com/onevcat/Kingfisher/blob/master/Sources/ImageCache.swift#L132) pointed to a path like Document, which will not be deleted by system. However, I am not sure whether it could pass the AppStore review, since there is a rule says that if the resource is downloadable and not created by users, it should be placed to Cache.

P.S. In fact, it is almost an abusing of `ImagePrefetcher` as well as `ImageCache`. Kingfisher is not a framework for file downloading after all. :(

**Comment by techyowl** [comment]
Thank you for getting back to me so quickly, I have setup a custom instance of an ImageCache object with a custom diskCachePathClosure option as you suggested and I'm still testing it. I did want to clarify the user will be prompted to cache the images locally and it is designed for offline access to Google Photos and I'm not certain it will be accepted by Apple but I kind of just wanted to build it because I needed it to work. I know this library isn't intended to download files but It is the most full featured caching library I know of for swift so I thought I could make it work. The user will be in control of which "Albums" are cached or not cached and will be able to manage the cache themselves. There were some apps that did this in the past but the developers have removed them from sale. Possible Apple Guideline violation like you said. Thanks for pointing me in the right direction though definitely helps. I'll re open this if I find it isn't the disk cache location and there might be a problem with the retrieve image from disk function until then thank you.

**Comment by techyowl** [comment]
I think i may be calling retrieve image in disk cache incorrectly. this is my current implementation:

`imagecache?.retrieveImageInDiskCache(forKey: img)`
where image cache is a custom instance of ImageCache with diskCachePathClosure defined. however I'm still having issues with nil images. I know the urls aren't different and I'm not sure why they would be nil if I'm prefetching them.

**Comment by onevcat** [comment]
Please remember that you need to create the prefetcher with the newly created cache like this:

```swift
ImagePrefetcher(urls: urls, options: [.targetCache(yourCache)])
```

And how about just 1 image? Does this only happens when there are lots of images? Could you build a minimal sample which could reproduce it? So I can have a deep look into it.

Thanks.

**Comment by techyowl** [comment]
I think I traced it down to networking issues. I was implementing it exactly like that and I noticed I can check for failedResources which I'm assuming is an array in the ImagePrefetcher and if so iterate through the prefetched until their isn't a failure or a failure of a low percentage or some sort of condition that will obtain the images when networking is restored. I'll work through the logic in a couple days. It seems to download small amount of images fine and I've had issues with the communication with google when requesting several hundred items instantaneously. I think I will do a 3 time retry as that is typically standard procedure when networking errors occur. I might also add a delay in there too because it can prefetch ~500 images no problem. I'll need to rework my code. thanks for clarifying I setup my image prefetched correctly.
